### PR TITLE
464059 Total bill 4

### PR DESCRIPTION
**Pull request by johnbenjamin-defra**

### CODE REVIEW

File: src/EPR.Calculator.API/Builder/Summary/Common/CalcResultSummaryUtil.cs [hunk 1]
```diff
@@ -818,4 +796,25 @@ public static decimal GetCommsCostHeaderWithBadDebtFor2bTitle(CalcResult calcRes
...
     }
+
+    public static decimal GetCountryOnePlusFourApportionment(CalcResult calcResult, Countries country)
+    {
+        var onePlusFourApportionment = calcResult.CalcResultOnePlusFourApportionment
+            .CalcResultOnePlusFourApportionmentDetails
+            .Single(x => x.Name == CalcResultCommsCostBuilder.OnePlusFourApportionment);
+
+        switch (country)
+        {
```
Comment (udaydefra) [inline src/EPR.Calculator.API/Builder/Summary/Common/CalcResultSummaryUtil.cs:807]: Default case and return 0

File: src/EPR.Calculator.API/Builder/Summary/CalcResultSummaryBuilder.cs [hunk 2]
```diff
@@ -374,13 +398,6 @@ private CalcResultSummaryProducerDisposalFees GetProducerRow(
...
 
-        private static decimal GetTotalOnePlus2AFeeWithBadDebtProvision(Dictionary<MaterialDetail, CalcResultSummaryProducerDisposalFeesByMaterial> materials, Dictionary<MaterialDetail, CalcResultSummaryProducerCommsFeesCostByMaterial> costSummary)
-        {
-            var totalLaDisposalFee = GetTotalProducerDisposalFeeWithBadDebtProvision(materials);
-            var commsCostBadDebt = GetTotalProducerCommsFeeWithBadDebtProvision(costSummary);
-            return totalLaDisposalFee + commsCostBadDebt;
-        }
-
         //section bad debt total
         public static decimal GetTotal1Plus2ABadDebt(List<MaterialDetail> materials, CalcResult calcResult)
```
Comment (udaydefra) [inline src/EPR.Calculator.API/Builder/Summary/CalcResultSummaryBuilder.cs:402]: Please use IEnumerable<MetarialDetail> unless you are modifying the list

File: src/EPR.Calculator.API/Builder/Summary/Common/CalcResultSummaryUtil.cs [hunk 3]
```diff
@@ -458,35 +418,45 @@ public static void SetHeaders(CalcResultSummary result, List<MaterialDetail> mat
 
     public static List<CalcResultSummaryHeader> GetProducerDisposalFeesHeaders()
     {
```
Comment (udaydefra) [inline src/EPR.Calculator.API/Builder/Summary/Common/CalcResultSummaryUtil.cs:420]: Please return IEnumerable<CalcResultSummaryHeader>

Comment (johnbenjamin-defra) [reply src/EPR.Calculator.API/Builder/Summary/Common/CalcResultSummaryUtil.cs:420]: I have applied the fix, thank you.

File: src/EPR.Calculator.API/Builder/Summary/CalcResultSummaryBuilder.cs (hunk 2 above)
Comment (johnbenjamin-defra) [reply src/EPR.Calculator.API/Builder/Summary/CalcResultSummaryBuilder.cs]: I have applied the fix

File: src/EPR.Calculator.API/Builder/Summary/Common/CalcResultSummaryUtil.cs (hunk 1 above)
Comment (johnbenjamin-defra) [reply src/EPR.Calculator.API/Builder/Summary/Common/CalcResultSummaryUtil.cs:807]: Thank you Uday, I have fixed the code now

File: src/EPR.Calculator.API/Builder/Summary/SaSetupCosts/SaSetupCostsProducer.cs [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        public static decimal GetProducerOneOffFeeWithBadDebtProvisionTotal(IEnumerable<ProducerDetail> producers,
+            IEnumerable<ProducerDetail> producersAndSubsidiaries,
+            IEnumerable<MaterialDetail> materials,
+            CalcResult calcResult)
+        {
+            return GetProducerOneOffFeeWithoutBadDebtProvisionTotal(producers, producersAndSubsidiaries, materials, calcResult) +
+                GetBadDebtProvisionTotal(producers, producersAndSubsidiaries, materials, calcResult);
+        }
+
+        public static decimal GetEnglandTotalWithBadDebtProvision(IEnumerable<ProducerDetail> producers,
```
Comment (tim-powell-eviden) [inline src/EPR.Calculator.API/Builder/Summary/SaSetupCosts/SaSetupCostsProducer.cs:98]: There's a lot of duplicated code here - could it be refactored to reduce duplication?

Comment (johnbenjamin-defra) [reply src/EPR.Calculator.API/Builder/Summary/SaSetupCosts/SaSetupCostsProducer.cs:98]: Thank you @tim-powell-eviden, as discussed I will leave it for now as the functions contain one-liner return statements and we will further improve the code later if required.

File: src/EPR.Calculator.API.UnitTests/Builder/Summary/LaDataPrepCosts/LaDataPrepCostsSummaryTests.cs [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+            {
+                CalcResultParameterOtherCost = new CalcResultParameterOtherCost
+                {
+                    BadDebtProvision = new KeyValuePair<string, string>("key1", "6%"),
+                    BadDebtValue = 6m,
+                    Details = [
+                        new CalcResultParameterOtherCostDetail {
+                            Name = "4 LA Data Prep Charge",
+                            OrderId = 1,
+                            England = "�40.00",
```
[the remark this answers is not in the document]
Comment (johnbenjamin-defra) [reply src/EPR.Calculator.API.UnitTests/Builder/Summary/LaDataPrepCosts/LaDataPrepCostsSummaryTests.cs:25]: Thank you @Mazar-Shaik, I have applied the fix now.

**Review by udaydefra — DISMISSED** [review]

**Review by Mazar-Shaik — APPROVED** [review]

**Review by udaydefra — APPROVED** [review]

**Review by Mazar-Shaik — APPROVED** [review]

**Pull request merged**